lib/socketpair.c

```diff
@@ -0,0 +1,88 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 2019, Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.haxx.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+#include "socketpair.h"
+
+#ifdef WIN32
+/*
+ * This is a socketpair() implementation for Windows.
+ */
+#include <string.h>
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#include <windows.h>
+#include <io.h>
+
+int Curl_socketpair(int domain, int type, int protocol,
+                    curl_socket_t socks[2])
+{
+  union {
+    struct sockaddr_in inaddr;
+    struct sockaddr addr;
+  } a;
+  curl_socket_t listener;
+  socklen_t addrlen = sizeof(a.inaddr);
+  int reuse = 1;
+  (void)domain;
+  (void)type;
+  (void)protocol;
+
+  listener = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+  if(listener == CURL_SOCKET_BAD)
+    return -1;
+
+  memset(&a, 0, sizeof(a));
+  a.inaddr.sin_family = AF_INET;
+  a.inaddr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+  a.inaddr.sin_port = 0;
+
+  socks[0] = socks[1] = CURL_SOCKET_BAD;
+
+  if(setsockopt(listener, SOL_SOCKET, SO_REUSEADDR,
+                (char *)&reuse, (socklen_t)sizeof(reuse)) == -1)
+    goto error;
+  if(bind(listener, &a.addr, sizeof(a.inaddr)) == -1)
+    goto error;
+  if(getsockname(listener, &a.addr, &addrlen) == -1)
+    goto error;
+  if(listen(listener, 1) == -1)
+    goto error;
+  socks[0] = WSASocketW(AF_INET, SOCK_STREAM, 0, NULL, 0, 0);
+  if(socks[0] == CURL_SOCKET_BAD)
+    goto error;
+  if(connect(socks[0], &a.addr, sizeof(a.inaddr)) == -1)
```

Comment:
Will probably block since accept() is below? curlx_nonblock() can help.

Reply:
Since this passes a lot of tests it seems unlikely.  I also got a pointer to [librdkafka's implementation](https://github.com/edenhill/librdkafka/blob/3cc2db5ae5177604223748cd92dbc9089d8a2e40/src/rdwin32.h#L242) of the same thing and they don't set it to non-blocking either.

Even though I suppose setting it non-blocking shouldn't hurt either and might make it work better on other platforms.

Reply:
I can't say for windows. On OS/400, curlx_nonblock() at https://github.com/curl/curl/blob/8bb3a95ce1577e2c8282e7d1c2830bde87cf5c3b/lib/vtls/gskit.c#L563 was mandatory.

Reply:
Are you really sure? Because I can build and use this version as-is on Linux and it works fine. There should be no need for non-blocking sockets here. It makes the code more complicated than without so I rather avoid it if I can...

Reply:
The connect() call blocks even on Linux if the listen queue is full.
That means an external process may DoS us by filling the queue between our listen() and connect() calls.
Although very improbable, this can occur.

Reply:
Ok, but for that to happen the attacker needs to run a program on the same host. Can and should we really protect against deliberate attacks from hostile programs running on the same host? What if that program allocates all the memory, then libcurl's malloc() calls will start to fail as well...

Reply:
The above mentioned socketpair implementation for windows doesn't protect against that. Neither do the two implementations [mentioned by Rene on the mailing list](https://curl.haxx.se/mail/lib-2019-10/0010.html).

Reply:
The current code already checks that the connection was the correct one.

Reply:
Sorry, I did not see it.
We could retry accept() if not verified, to avoid a failure upon attack?

Reply:
Yes, that makes sense!

Reply:
Wait, did I do that check the wrong way around? Hm...

Reply:
Yes, I think so :-(

Reply:
Not only that. When I try this address verifying code on Linux, the `sin_port` fields never match even when no evil is going on...!
~~~
(gdb) n
87          if((a.inaddr.sin_addr.s_addr == b.inaddr.sin_addr.s_addr) &&
(gdb) list
82        do {
83          socks[1] = accept(listener, &b.addr, &blen);
84          if(socks[1] == CURL_SOCKET_BAD)
85            goto error;
86          /* verify the accepted connection */
87          if((a.inaddr.sin_addr.s_addr == b.inaddr.sin_addr.s_addr) &&
88             (a.inaddr.sin_port == b.inaddr.sin_port))
89            break;
90        } while(1);
91
(gdb) p a.inaddr.sin_port
$1 = 62363
(gdb) p b.inaddr.sin_port
$2 = 53928

Reply:
I opted on a more custom verification method in my new take.

Reply:
> Not only that. When I try this address verifying code on Linux, the sin_port fields never match even when no evil is going on...!

Did you preset `blen` to `sizeof(b.addr)` `before accept()`?

Reply:
It was set once and only used in that call and the same size as `addrlen`, so that wasn't a problem